File: src/annotation/test/TestModelHistory.c
/**
 * \file    TestModelHistory.cpp
 * \brief   ModelHistory unit tests
 * \author  Sarah Keating
 *
 * $Id$
 * $Source$
 */
/* Copyright 2007 California Institute of Technology.
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation.  A copy of the license agreement is
 * provided in the file named "LICENSE.txt" included with this software
 * distribution.  It is also available online at
 * http://sbml.org/software/libsbml/license.html
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include "common/common.h"
#include "sbml/common/extern.h"
#include "util/List.h"
#include "ModelHistory.h"
#include "../../xml/XMLNode.h"


#include <check.h>



START_TEST (test_Date_create)
{
  Date_t * date = Date_createFromValues(2005, 12, 30, 12, 15, 45, 1, 2, 0);

  fail_unless(date != NULL);
  fail_unless(Date_getYear(date) == 2005);
  fail_unless(Date_getMonth(date) == 12);
  fail_unless(Date_getDay(date) == 30);
  fail_unless(Date_getHour(date) == 12);
  fail_unless(Date_getMinute(date) == 15);
  fail_unless(Date_getSecond(date) == 45);
  fail_unless(Date_getSignOffset(date) == 1);
  fail_unless(Date_getHoursOffset(date) == 2);
  fail_unless(Date_getMinutesOffset(date) == 0);

  Date_free(date);
}
END_TEST

START_TEST (test_Date_createFromString)
{
  const char * dd = "2012-12-02T14:56:11Z";

  Date_t * date = Date_createFromString(dd);

  fail_unless(date != NULL);
  fail_unless(!strcmp(Date_getDateAsString(date), "2012-12-02T14:56:11Z"));
  fail_unless(Date_getYear(date) == 2012);
  fail_unless(Date_getMonth(date) == 12);
  fail_unless(Date_getDay(date) == 2);
  fail_unless(Date_getHour(date) == 14);
  fail_unless(Date_getMinute(date) == 56);
  fail_unless(Date_getSecond(date) == 11);
  fail_unless(Date_getSignOffset(date) == 0);
  fail_unless(Date_getHoursOffset(date) == 0);
  fail_unless(Date_getMinutesOffset(date) == 0);

  Date_free(date);
}
END_TEST

START_TEST (test_Date_setters)
{
  Date_t * date = Date_createFromValues(2005, 12, 30, 12, 15, 45, 1, 2, 0);
  fail_unless(date != NULL);

  Date_setYear(date, 2012);
  Date_setMonth(date, 3);
  Date_setDay(date, 28);
  Date_setHour(date, 23);
  Date_setMinute(date, 4);
  Date_setSecond(date, 32);
  Date_setSignOffset(date, 1);
  Date_setHoursOffset(date, 2);
  Date_setMinutesOffset(date, 32);

  fail_unless(Date_getYear(date) == 2012);
  fail_unless(Date_getMonth(date) == 3);
  fail_unless(Date_getDay(date) == 28);
  fail_unless(Date_getHour(date) == 23);
  fail_unless(Date_getMinute(date) == 4);
  fail_unless(Date_getSecond(date) == 32);
  fail_unless(Date_getSignOffset(date) == 1);
  fail_unless(Date_getHoursOffset(date) == 2);
  fail_unless(Date_getMinutesOffset(date) == 32);
  fail_unless(!strcmp(Date_getDateAsString(date), "2012-03-28T23:04:32+02:32"));

  Date_free(date);
}
END_TEST

START_TEST (test_Date_getDateAsString)
{
  char * dd = "2005-02-02T14:56:11Z";

  Date_t * date = Date_createFromString(dd);

  fail_unless(date != NULL);
  fail_unless(Date_getYear(date) == 2005);
  fail_unless(Date_getMonth(date) == 2);
  fail_unless(Date_getDay(date) == 2);
  fail_unless(Date_getHour(date) == 14);
  fail_unless(Date_getMinute(date) == 56);
  fail_unless(Date_getSecond(date) == 11);
  fail_unless(Date_getSignOffset(date) == 0);
  fail_unless(Date_getHoursOffset(date) == 0);
  fail_unless(Date_getMinutesOffset(date) == 0);

  Date_setYear(date, 2012);
  Date_setMonth(date, 3);
  Date_setDay(date, 28);
  Date_setHour(date, 23);
  Date_setMinute(date, 4);
  Date_setSecond(date, 32);
  Date_setSignOffset(date, 1);
  Date_setHoursOffset(date, 2);
  Date_setMinutesOffset(date, 32);

  fail_unless(!strcmp(Date_getDateAsString(date), "2012-03-28T23:04:32+02:32"));

  Date_free(date);
}
END_TEST

START_TEST(test_ModelCreator_create)
{
  ModelCreator_t * mc = ModelCreator_create();

  fail_unless(mc != NULL);

  ModelCreator_free(mc);

}
END_TEST


START_TEST(test_ModelCreator_setters)
{
  ModelCreator_t * mc = ModelCreator_create();

  fail_unless(mc != NULL);

  fail_unless(ModelCreator_isSetFamilyName(mc) == 0);
  fail_unless(ModelCreator_isSetGivenName(mc) == 0);
  fail_unless(ModelCreator_isSetEmail(mc) == 0);
  fail_unless(ModelCreator_isSetOrganisation(mc) == 0);

  ModelCreator_setFamilyName(mc, "Keating");
  ModelCreator_setGivenName(mc, "Sarah");
  ModelCreator_setEmail(mc, "sbml-team@caltech.edu");
  ModelCreator_setOrganisation(mc, "UH");

  fail_unless(!strcmp(ModelCreator_getFamilyName(mc), "Keating"));
  fail_unless(!strcmp(ModelCreator_getGivenName(mc), "Sarah"));
  fail_unless(!strcmp(ModelCreator_getEmail(mc), "sbml-team@caltech.edu"));
  fail_unless(!strcmp(ModelCreator_getOrganisation(mc), "UH"));

  fail_unless(ModelCreator_isSetFamilyName(mc) == 1);
  fail_unless(ModelCreator_isSetGivenName(mc) == 1);
  fail_unless(ModelCreator_isSetEmail(mc) == 1);
  fail_unless(ModelCreator_isSetOrganisation(mc) == 1);

  ModelCreator_unsetFamilyName(mc);
  ModelCreator_unsetGivenName(mc);
  ModelCreator_unsetEmail(mc);
  ModelCreator_unsetOrganisation(mc);

  fail_unless(!strcmp(ModelCreator_getFamilyName(mc), ""));
  fail_unless(!strcmp(ModelCreator_getGivenName(mc), ""));
  fail_unless(!strcmp(ModelCreator_getEmail(mc), ""));
  fail_unless(!strcmp(ModelCreator_getOrganisation(mc), ""));

  fail_unless(ModelCreator_isSetFamilyName(mc) == 0);
  fail_unless(ModelCreator_isSetGivenName(mc) == 0);
  fail_unless(ModelCreator_isSetEmail(mc) == 0);
  fail_unless(ModelCreator_isSetOrganisation(mc) == 0);
  ModelCreator_free(mc);

}
END_TEST

START_TEST (test_ModelHistory_create)
{
  ModelHistory_t * history = ModelHistory_create();

  fail_unless(history != NULL);
  fail_unless(ModelHistory_getListCreators(history) != NULL);
  fail_unless(ModelHistory_getCreatedDate(history) == NULL);
  fail_unless(ModelHistory_getModifiedDate(history) == NULL);

  ModelHistory_free(history);
}
END_TEST

START_TEST (test_ModelHistory_addCreator)
{
  ModelCreator_t * newMC;
  ModelHistory_t * history = ModelHistory_create();

  fail_unless(ModelHistory_getNumCreators(history) == 0);

  fail_unless(history != NULL);

  ModelCreator_t * mc = ModelCreator_create();
  fail_unless(mc != NULL);

  ModelCreator_setFamilyName(mc, "Keating");
  ModelCreator_setGivenName(mc, "Sarah");
  ModelCreator_setEmail(mc, "sbml-team@caltech.edu");
  ModelCreator_setOrganisation(mc, "UH");

  ModelHistory_addCreator(history, mc);

  fail_unless(ModelHistory_getNumCreators(history) == 1);
  ModelCreator_free(mc);

  newMC = List_get(ModelHistory_getListCreators(history), 0);
  fail_unless(newMC != NULL);

  fail_unless(!strcmp(ModelCreator_getFamilyName(newMC), "Keating"));
  fail_unless(!strcmp(ModelCreator_getGivenName(newMC), "Sarah"));
  fail_unless(!strcmp(ModelCreator_getEmail(newMC), "sbml-team@caltech.edu"));
  fail_unless(!strcmp(ModelCreator_getOrganisation(newMC), "UH"));

  ModelCreator_free(newMC);

  ModelHistory_free(history);
}
END_TEST

START_TEST (test_ModelHistory_setCreatedDate)
{
  ModelHistory_t * history = ModelHistory_create();

  fail_unless(history != NULL);

  fail_unless(ModelHistory_isSetCreatedDate(history) == 0);

  Date_t * date = Date_createFromValues(2005, 12, 30, 12, 15, 45, 1, 2, 0);
  
  ModelHistory_setCreatedDate(history, date);
  fail_unless(ModelHistory_isSetCreatedDate(history) == 1);

  Date_free(date);

  Date_t * newdate = ModelHistory_getCreatedDate(history);
  fail_unless(Date_getYear(newdate) == 2005);
  fail_unless(Date_getMonth(newdate) == 12);
  fail_unless(Date_getDay(newdate) == 30);
  fail_unless(Date_getHour(newdate) == 12);
  fail_unless(Date_getMinute(newdate) == 15);
  fail_unless(Date_getSecond(newdate) == 45);
  fail_unless(Date_getSignOffset(newdate) == 1);
  fail_unless(Date_getHoursOffset(newdate) == 2);
  fail_unless(Date_getMinutesOffset(newdate) == 0);

  ModelHistory_free(history);

}
END_TEST


START_TEST (test_ModelHistory_setModifiedDate)
{
  ModelHistory_t * history = ModelHistory_create();

  fail_unless(history != NULL);
  fail_unless(ModelHistory_isSetModifiedDate(history) == 0);

  Date_t * date = Date_createFromValues(2005, 12, 30, 12, 15, 45, 1, 2, 0);
  
  ModelHistory_setModifiedDate(history, date);
//  Date_free(date);
  fail_unless(ModelHistory_isSetModifiedDate(history) == 1);

  Date_t * newdate = ModelHistory_getModifiedDate(history);
  fail_unless(Date_getYear(newdate) == 2005);
  fail_unless(Date_getMonth(newdate) == 12);
  fail_unless(Date_getDay(newdate) == 30);
  fail_unless(Date_getHour(newdate) == 12);
  fail_unless(Date_getMinute(newdate) == 15);
  fail_unless(Date_getSecond(newdate) == 45);
  fail_unless(Date_getSignOffset(newdate) == 1);
  fail_unless(Date_getHoursOffset(newdate) == 2);
  fail_unless(Date_getMinutesOffset(newdate) == 0);

  ModelHistory_free(history);
}
END_TEST


Suite *
create_suite_ModelHistory (void)
{
  Suite *suite = suite_create("ModelHistory");
  TCase *tcase = tcase_create("ModelHistory");


  tcase_add_test( tcase, test_Date_create  );
  tcase_add_test( tcase, test_Date_createFromString  );
  tcase_add_test( tcase, test_Date_setters  );
  tcase_add_test( tcase, test_Date_getDateAsString  );
  tcase_add_test( tcase, test_ModelCreator_create  );
  tcase_add_test( tcase, test_ModelCreator_setters  );
  tcase_add_test( tcase, test_ModelHistory_create  );
  tcase_add_test( tcase, test_ModelHistory_addCreator  );
  tcase_add_test( tcase, test_ModelHistory_setCreatedDate  );
  tcase_add_test( tcase, test_ModelHistory_setModifiedDate  );
  suite_add_tcase(suite, tcase);

  return suite;
}


END_C_DECLS

File: src/bindings/python/test/annotation/TestModelHistory.py
#
# @file    TestModelHistory.py
# @brief   ModelHistory unit tests
#
# @author  Akiya Jouraku (Python conversion)
# @author  Sarah Keating 
#
# $Id:$
# $HeadURL:$
#
# This test file was converted from src/sbml/test/TestModelHistory.c
# with the help of conversion sciprt (ctest_converter.pl).
#
#<!---------------------------------------------------------------------------
# This file is part of libSBML.  Please visit http://sbml.org for more
# information about SBML, and the latest version of libSBML.
#
# Copyright 2005-2008 California Institute of Technology.
# Copyright 2002-2005 California Institute of Technology and
#                     Japan Science and Technology Corporation.
# 
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation.  A copy of the license agreement is provided
# in the file named "LICENSE.txt" included with this software distribution
# and also available online as http://sbml.org/software/libsbml/license.html
#--------------------------------------------------------------------------->*/
import sys
import unittest
import libsbml

class TestModelHistory(unittest.TestCase):


  def test_Date_create(self):
    date = libsbml.Date(2005,12,30,12,15,45,1,2,0)
    self.assert_( date != None )
    self.assert_( date.getYear() == 2005 )
    self.assert_( date.getMonth() == 12 )
    self.assert_( date.getDay() == 30 )
    self.assert_( date.getHour() == 12 )
    self.assert_( date.getMinute() == 15 )
    self.assert_( date.getSecond() == 45 )
    self.assert_( date.getSignOffset() == 1 )
    self.assert_( date.getHoursOffset() == 2 )
    self.assert_( date.getMinutesOffset() == 0 )
    date = None
    pass  

  def test_Date_createFromString(self):
    dd =  "2012-12-02T14:56:11Z";
    date = libsbml.Date(dd)
    self.assert_( date != None )
    self.assert_((  "2012-12-02T14:56:11Z" == date.getDateAsString() ))
    self.assert_( date.getYear() == 2012 )
    self.assert_( date.getMonth() == 12 )
    self.assert_( date.getDay() == 2 )
    self.assert_( date.getHour() == 14 )
    self.assert_( date.getMinute() == 56 )
    self.assert_( date.getSecond() == 11 )
    self.assert_( date.getSignOffset() == 0 )
    self.assert_( date.getHoursOffset() == 0 )
    self.assert_( date.getMinutesOffset() == 0 )
    date = None
    pass  

  def test_Date_getDateAsString(self):
    dd =  "2005-02-02T14:56:11Z";
    date = libsbml.Date(dd)
    self.assert_( date != None )
    self.assert_( date.getYear() == 2005 )
    self.assert_( date.getMonth() == 2 )
    self.assert_( date.getDay() == 2 )
    self.assert_( date.getHour() == 14 )
    self.assert_( date.getMinute() == 56 )
    self.assert_( date.getSecond() == 11 )
    self.assert_( date.getSignOffset() == 0 )
    self.assert_( date.getHoursOffset() == 0 )
    self.assert_( date.getMinutesOffset() == 0 )
    date.setYear(2012)
    date.setMonth(3)
    date.setDay(28)
    date.setHour(23)
    date.setMinute(4)
    date.setSecond(32)
    date.setSignOffset(1)
    date.setHoursOffset(2)
    date.setMinutesOffset(32)
    self.assert_((  "2012-03-28T23:04:32+02:32" == date.getDateAsString() ))
    date = None
    pass  

  def test_Date_setters(self):
    date = libsbml.Date(2005,12,30,12,15,45,1,2,0)
    self.assert_( date != None )
    date.setYear(2012)
    date.setMonth(3)
    date.setDay(28)
    date.setHour(23)
    date.setMinute(4)
    date.setSecond(32)
    date.setSignOffset(1)
    date.setHoursOffset(2)
    date.setMinutesOffset(32)
    self.assert_( date.getYear() == 2012 )
    self.assert_( date.getMonth() == 3 )
    self.assert_( date.getDay() == 28 )
    self.assert_( date.getHour() == 23 )
    self.assert_( date.getMinute() == 4 )
    self.assert_( date.getSecond() == 32 )
    self.assert_( date.getSignOffset() == 1 )
    self.assert_( date.getHoursOffset() == 2 )
    self.assert_( date.getMinutesOffset() == 32 )
    self.assert_((  "2012-03-28T23:04:32+02:32" == date.getDateAsString() ))
    date = None
    pass  

  def test_ModelCreator_create(self):
    mc = libsbml.ModelCreator()
    self.assert_( mc != None )
    mc = None
    pass  

  def test_ModelCreator_setters(self):
    mc = libsbml.ModelCreator()
    self.assert_( mc != None )
    self.assert_( mc.isSetFamilyName() == 0 )
    self.assert_( mc.isSetGivenName() == 0 )
    self.assert_( mc.isSetEmail() == 0 )
    self.assert_( mc.isSetOrganisation() == 0 )
    mc.setFamilyName( "Keating")
    mc.setGivenName( "Sarah")
    mc.setEmail( "sbml-team@caltech.edu")
    mc.setOrganisation( "UH")
    self.assert_((  "Keating" == mc.getFamilyName() ))
    self.assert_((  "Sarah" == mc.getGivenName() ))
    self.assert_((  "sbml-team@caltech.edu" == mc.getEmail() ))
    self.assert_((  "UH" == mc.getOrganisation() ))
    self.assert_( mc.isSetFamilyName() == 1 )
    self.assert_( mc.isSetGivenName() == 1 )
    self.assert_( mc.isSetEmail() == 1 )
    self.assert_( mc.isSetOrganisation() == 1 )
    mc.unsetFamilyName()
    mc.unsetGivenName()
    mc.unsetEmail()
    mc.unsetOrganisation()
    self.assert_((  "" == mc.getFamilyName() ))
    self.assert_((  "" == mc.getGivenName() ))
    self.assert_((  "" == mc.getEmail() ))
    self.assert_((  "" == mc.getOrganisation() ))
    self.assert_( mc.isSetFamilyName() == 0 )
    self.assert_( mc.isSetGivenName() == 0 )
    self.assert_( mc.isSetEmail() == 0 )
    self.assert_( mc.isSetOrganisation() == 0 )
    self.assert_( mc.isSetOrganization() == 0 )
    mc.setOrganization( "UH")
    self.assert_((  "UH" == mc.getOrganization() ))
    self.assert_( mc.isSetOrganization() == 1 )
    mc.unsetOrganisation()
    self.assert_((  "" == mc.getOrganization() ))
    self.assert_( mc.isSetOrganization() == 0 )
    mc = None
    pass  

  def test_ModelHistory_addCreator(self):
    history = libsbml.ModelHistory()
    self.assert_( history.getNumCreators() == 0 )
    self.assert_( history != None )
    mc = libsbml.ModelCreator()
    self.assert_( mc != None )
    mc.setFamilyName( "Keating")
    mc.setGivenName( "Sarah")
    mc.setEmail( "sbml-team@caltech.edu")
    mc.setOrganisation( "UH")
    history.addCreator(mc)
    self.assert_( history.getNumCreators() == 1 )
    mc = None
    newMC = history.getCreator(0)
    self.assert_( newMC != None )
    self.assert_((  "Keating" == newMC.getFamilyName() ))
    self.assert_((  "Sarah" == newMC.getGivenName() ))
    self.assert_((  "sbml-team@caltech.edu" == newMC.getEmail() ))
    self.assert_((  "UH" == newMC.getOrganisation() ))
    history = None
    pass  

  def test_ModelHistory_create(self):
    history = libsbml.ModelHistory()
    self.assert_( history != None )
    self.assert_( history.getCreatedDate() == None )
    self.assert_( history.getModifiedDate() == None )
    history = None
    pass  

  def test_ModelHistory_setCreatedDate(self):
    history = libsbml.ModelHistory()
    self.assert_( history != None )
    self.assert_( history.isSetCreatedDate() == 0 )
    date = libsbml.Date(2005,12,30,12,15,45,1,2,0)
    history.setCreatedDate(date)
    self.assert_( history.isSetCreatedDate() == 1 )
    date = None
    newdate = history.getCreatedDate()
    self.assert_( newdate.getYear() == 2005 )
    self.assert_( newdate.getMonth() == 12 )
    self.assert_( newdate.getDay() == 30 )
    self.assert_( newdate.getHour() == 12 )
    self.assert_( newdate.getMinute() == 15 )
    self.assert_( newdate.getSecond() == 45 )
    self.assert_( newdate.getSignOffset() == 1 )
    self.assert_( newdate.getHoursOffset() == 2 )
    self.assert_( newdate.getMinutesOffset() == 0 )
    history = None
    pass  

  def test_ModelHistory_setModifiedDate(self):
    history = libsbml.ModelHistory()
    self.assert_( history != None )
    self.assert_( history.isSetModifiedDate() == 0 )
    date = libsbml.Date(2005,12,30,12,15,45,1,2,0)
    history.setModifiedDate(date)
    date = None
    self.assert_( history.isSetModifiedDate() == 1 )
    newdate = history.getModifiedDate()
    self.assert_( newdate.getYear() == 2005 )
    self.assert_( newdate.getMonth() == 12 )
    self.assert_( newdate.getDay() == 30 )
    self.assert_( newdate.getHour() == 12 )
    self.assert_( newdate.getMinute() == 15 )
    self.assert_( newdate.getSecond() == 45 )
    self.assert_( newdate.getSignOffset() == 1 )
    self.assert_( newdate.getHoursOffset() == 2 )
    self.assert_( newdate.getMinutesOffset() == 0 )
    history = None
    pass  

def suite():
  suite = unittest.TestSuite()
  suite.addTest(unittest.makeSuite(TestModelHistory))

  return suite

if __name__ == "__main__":
  if unittest.TextTestRunner(verbosity=1).run(suite()).wasSuccessful() :
    sys.exit(0)
  else:
    sys.exit(1)
